enhancements/update/standalone-worker-surge.md

```diff
@@ -0,0 +1,295 @@
+stalled drain to succeed -- helping to ensure that updates can proceed (by incurring disruption) even 
+in the presence of poorly configured workloads.
+
+## Motivation
+
+During a typical worker node update for an OpenShift cluster, it is necessary to "cordon" nodes (prevent new pods from being scheduled on a node)
+and "drain" them (attempt to migrate workloads by rescheduling its pods onto uncordoned nodes). Workers generally need to be rebooted during a
+cluster update and draining nodes is best practice before rebooting them. If they were not drained first, pods running
+on a node targeted by the update process could be terminated with no other viable pods on the cluster to 
+handle the workload. This outcome can cause a disruption in the service the terminated pod was attempt to provide. For example,
+an incoming web request may not be routable to a pod for a given Kubernetes service - resulting in errors being returned
+to the consumers of that service.
+
+With appropriate cluster management, node draining can be used to ensure that sufficient pods are running to satisfy workload requirements
+at all times - even during updates. "Appropriate cluster management," though, is a multi-faceted challenge involving considerations
+from pod replicas to cluster topology.
+
+### Managing Worker Node Capacity
+
+One aspect of this challenge is ensuring that, while a node is being drained, there is sufficient worker node capacity (CPU/memory/other
+resources) available for new pods to take the place of old pods from the node being drained. Consider the reductive example of 
+a static cluster with a single worker node. If there is an attempt to drain the node in this example, there is no additional worker 
+node capacity available to schedule new pods to replace the pods being drained. This can result in a stalled drain -- one that 
+does not terminate until there is external intervention.
+
+Stalled drains create a frustrating experience for operations teams as they require analysis and intervention. They can also
+make it impossible to predict when an update will complete -- complicating work schedules and communications. There are a number of reasons 
+drains can stall, but simple lack of spare worker node capacity is a common one. The reason is that running an over-provisioned cluster
+(i.e. with more worker nodes than actually required for the steady state workloads) is not always cost-effective. Running
+an extra node 24x7 simply to ensure capacity for the short period while another node is being drained is inefficient.
+
+One cost-effective approach to ensure capacity is called "surging". With a surge strategy, during an update, the platform
+is permitted to bring additional worker nodes online, to accommodate workloads being drained from existing nodes. After an 
+update concludes, the surged nodes are scaled down and the cluster resumes its steady state.

```

Comment:
In OSD/ROSA we scale all machinesets by creating a new machineset with replicas=1 that matches the original machineset in all but name (we append some string).  In this proposal I'd expect as a machineset is being updated the same happens, blindly scaling everything by the surge amount.

Agree on the disruption!  If it were possible to reduce that it would be great.  A challenge we have had for managed openshift to date is machineset is not immuable while upgrades happen, hence the new machineset creation to isolate from any changes that might come in while an upgrade is in progress.  This means we can't get away from this disruption.  If surge were a first class citizen and this tweak to simply destroy non-updated machines to get back to the non-surge replicas it would be very nice.

Reply:
I am adding implementation details after further discussion with lifecycle that mirror what OSD has done. This should address the machineset shape question directly.
For disruption, I don't think any update mechanism tries to prevent double disruption (of non-HA workloads). Unless the kube-scheduler can add weight to "newest nodes", there is every chance a drained workload will find a seat on a old node that is, itself, soon to be drained..

Reply:
> In OSD/ROSA we scale all machinesets by creating a new machineset with replicas=1 that matches the original machineset in all but name

You just mean during upgrades right? Not during normal operation?

> If surge were a first class citizen and this tweak to simply destroy non-updated machines to get back to the non-surge replicas it would be very nice.

I believe that is the behaviour of the Cluster API MachineDeployment already, do you know if that is the behaviour on HCP?

> For disruption, I don't think any update mechanism tries to prevent double disruption (of non-HA workloads). Unless the kube-scheduler can add weight to "newest nodes", there is every chance a drained workload will find a seat on a old node that is, itself, soon to be drained..

Yeah true, but I think the current suggestion of scaling up and then down again makes it more likely/perhaps opens us up to triple?

Reply:
Yes, only in upgrade.

In HCP we're using default "Replace" behavior.  I hope it's not updating a node just to delete one.  Asking to confirm, will follow up here.

I haven't heard a requirement to optimize for least disruption of workloads.  That feels like a larger thing to tackle and is not a new problem brought on by this enhancement.  Scheduling to a yet-to-be-updated node is going to happen today 100% of the time for the first node update with additional chances on subsequent nodes.

Reply:
> I haven't heard a requirement to optimize for least disruption of workloads. That feels like a larger thing to tackle and is not a new problem brought on by this enhancement. Scheduling to a yet-to-be-updated node is going to happen today 100% of the time for the first node update with additional chances on subsequent nodes.

Yes agreed, but, if we know that this is a problem and we are making choices we know make this worse, do we really want to make those choices? We don't need to fix it now, but I'd have thought we would not want to add something we _know_ makes the situation worse

Reply:
> I'd have thought we would not want to add something we know makes the situation worse

Would it be a considered a regression for this behavior to change and for restarts of workloads to happen more often?  If not, who makes the call to make that a concern we're willing to consider for regression?

Reply:
I think that depends on whether this is something we test. I know that we do test disruption of workloads during upgrades today. For example, there are load balancer tests with example workloads that test how much downtime a customer application would face as we go through an upgrade.
Currently, because that works with PDBs and the LB health checks for the various clouds, I think that this wouldn't necessarily fall foul of those.

But I'd be interested to see what the TRT folks think, since they are the ones who look after our testing and pick up regressions like this.